RNTester/NativeModuleExample/Screenshot.mm

```diff
@@ -0,0 +1,137 @@
+  CGRect imageBounds = CGRectNull;
+  CGImageRef windowImage = CGWindowListCreateImage(
+    imageBounds,
+    listOptions,
+    windowID,
+    imageOptions);
+  NSImage* image = [[NSImage alloc] initWithCGImage:windowImage size:[keyWindow frame].size];
+
+  // save to a temp file
+  NSError *error = nil;
+  NSString *tempFilePath = RCTTempFilePath(@"jpeg", &error);
+  NSData* imageData = [image TIFFRepresentation];
+  NSBitmapImageRep* imageRep = [NSBitmapImageRep imageRepWithData:imageData];
+  NSDictionary* imageProps =
+    [NSDictionary
+    dictionaryWithObject:@0.8
+    forKey:NSImageCompressionFactor
+    ];
+  imageData = [imageRep representationUsingType:NSBitmapImageFileTypeJPEG properties:imageProps];
+  [imageData writeToFile:tempFilePath atomically:NO];
+  
+  return tempFilePath;
+}
+
+class ScreenshotManagerTurboModule : public facebook::react::TurboModule
+{
+public:
+  ScreenshotManagerTurboModule(std::shared_ptr<facebook::react::CallInvoker> jsInvoker)
+    :facebook::react::TurboModule("ScreenshotManager", jsInvoker)
+  {
+  }
+  
+  facebook::jsi::Value get(
+    facebook::jsi::Runtime& runtime,
+    const facebook::jsi::PropNameID& propName
+  ) override
+  {
+    auto jsInvoker = jsInvoker_;
+    auto key = propName.utf8(runtime);
+    if (key == "takeScreenshot")
+    {
+      return facebook::jsi::Function::createFromHostFunction(
+        runtime,
+        propName,
+        0,
+        [jsInvoker](
+          facebook::jsi::Runtime& runtime,
+          const facebook::jsi::Value& thisVal,
+          const facebook::jsi::Value *args,
+          size_t count)
+        {
+          return facebook::react::createPromiseAsJSIValue(
+            runtime,
+            [jsInvoker](facebook::jsi::Runtime& runtime, std::shared_ptr<facebook::react::Promise> promise)
+            {
+              // ignore arguments, assume to be ('window', {format: 'jpeg', quality: 0.8})
+
+              dispatch_async(dispatch_get_main_queue(), ^{
+                NSString* tempFilePath = SaveScreenshotToTempFile();
+                jsInvoker->invokeAsync([tempFilePath, &runtime, promise]()
+                {
+                  promise->resolve(facebook::jsi::Value(
+                    runtime,
+                    facebook::jsi::String::createFromUtf8(
+                      runtime,
+                      std::string([tempFilePath UTF8String])
+                      )
+                    ));
+                });
+              });
+            }
+          );
+        }
+      );
+    }
+    else
+    {
+      return facebook::jsi::Value::undefined();
+    }
+  }
+};
+
+@implementation ScreenshotManagerTurboModuleManagerDelegate
+
+- (std::shared_ptr<facebook::react::TurboModule>)
+  getTurboModule:(const std::string &)name
+  jsInvoker:(std::shared_ptr<facebook::react::CallInvoker>)jsInvoker
+{
+  if (name == "ScreenshotManager")
+  {
+    return std::make_shared<ScreenshotManagerTurboModule>(jsInvoker);
+  }
+  return nullptr;
+}
```

Comment:
Actually it is the optional one that is important, because I am seeking to implement turbo modules in C++, not in objective C. The required one can be empty. This sounds wired but it works like this.

Reply:
Oh ok, that is a bit counter-intuitive indeed. Would be good to update the docs in the delegate protocol too at some point.